julia/src/native_function.jl

```diff
@@ -0,0 +1,277 @@
+
+
+
+"""
+    NativeFunction{R}
+
+Callable Julia wrapper of some native Dex function.
+Returns type from calling it is is `R`.
+
+Usually constructed using [`@dex_func_str`](@ref),
+or via `NativeFunction(atom)` on some [`DexCall.Atom`](@ref).
+"""
+struct NativeFunction{R} <: Function
+    c_func_ptr::Ptr{Nothing}
+    implict_argument_signature::Vector{Binder}
+    explicit_argument_signature::Vector{Binder}
+    result_signature::Vector{Binder}
+end
+
+NativeFunction(atom::Atom, jit=JIT) = NativeFunction(atom.ptr, atom.ctx, jit)
+function NativeFunction(atom::Ptr{HsAtom}, ctx=PRELUDE, jit=JIT)
+    c_func_ptr = compile(ctx, atom, jit)
+    sig_ptr = get_function_signature(c_func_ptr, jit)
+    sig_ptr == C_NULL && error("Failed to retrieve the function signature")
+
+    try
+        signature = NativeFunctionSignatureJL(sig_ptr)
+        argument_signature = parse_sig(signature.arg)
+        result_signature = parse_sig(signature.res)
+        R = result_type(result_signature)
+        f = NativeFunction{R}(
+            c_func_ptr,
+            filter(x->x.implicit, argument_signature),
+            filter(x->!x.implicit, argument_signature),
+            result_signature
+        )
+
+        # We are attaching this to f.result_signature because only mutable types can have finalizers
+        finalizer(f.result_signature) do _
+            unload(c_func_ptr, jit)
+        end
+        return f
+    catch
+        unload(c_func_ptr, jit)
+        rethrow()
+    finally
+        free_function_signature(sig_ptr)
+    end
+end
+
+function (f::NativeFunction{R})(args...)::R where R
+    if length(args) != length(f.explicit_argument_signature)
+        throw(ArgumentError("wrong number of inputs, expected: $(length(f.explicit_argument_signature) )"))
+    end
+
+    named_sizes = Dict{Symbol, Int}()
+    arg_ctypes = map(f.explicit_argument_signature, args) do binder, arg
+        to_ctype!(named_sizes, binder.type, arg)
+    end
+
+    implict_values = map(f.implict_argument_signature) do binder
+        # Right now the Exported API only supports implicts that are sizes
+        # and now all of them will have been determined by the preverious `to_ctype!` step
+        @assert binder.type == Int32
+        named_sizes[binder.name]
+    end
+
+    results_ptrs_and_thunks = map(f.result_signature) do builder
+        create(builder.type, named_sizes)  
+    end
+
+    result_ptrs = first.(results_ptrs_and_thunks)
+    result_thunks = Tuple(last.(results_ptrs_and_thunks))
+
+    args = (implict_values..., arg_ctypes..., result_ptrs...)
+    # No return value, that we care about because it write's its return into  the results_ptrs
+    _ccall_dex_func(f.c_func_ptr, args...)
+
+    if length(result_thunks) === 1
+        only(result_thunks)()
+    else
+        map(t->t(), result_thunks)
+    end
+end
+
+# Work around needing to have the return types as literals, by defining a generated function that does.
```

Comment:
Noone should be expected to follow what this code does.
Like even people who have been using julia for years shouldn't.
It's pretty gross.

The sensible version of this code is:
````julia
ccall(f_ptr, Int64, typeof.(args), args...)
```
What this code does is call a function that has a C API, with return type Int64, and with argument types that are exactly the types of the `args`
The reason it does it this way is because `ccall` is not a function but a built-in language intrinsic that is largely resolved at compile time, and that needs the type information given to it as literals.
`typeof.(args)` is not a literal which is a problem -- even though it can actually be resolved at compile-time -- the information exists.
So we use a generated function in order to get at the information.
Docs on those are at https://docs.julialang.org/en/v1/manual/metaprogramming/#Generated-functions
The short version though is that they take in arguments at compile time and they only get their types, then they return a AST which is what is run at runtime.
So what we are doing is building the AST that is for the ccall, which has the types all present as literals.


I will add a comment to the code to explain all this.
And an issue on JuliaLang/julia about how much this sucks.

Reply:
Thank you, that's super helpful.